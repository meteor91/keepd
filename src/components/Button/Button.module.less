@import "styles/colors.less";
@import "styles/paragraphs.less";

.btn {
  display: inline-flex;
  justify-content: center;
  align-items: center;
  background: @blue-500;
  border-radius: 6px;
  padding: 6px 12px;
  border: 0;
  color: @white;

  .paragraph-medium-4();

  &:active,
  &:hover {
    background: @blue-600;
  }

  &:focus {
    box-shadow: 0 1px 2px @blue-50, 0 0 0 3px @blue-50;
    border-radius: 6px;
  }

  &:disabled {
    background: @blue-100;
    cursor: not-allowed;
  }

  &-secondary {
    color: @blue-400;
    border: 1px solid @blue-400;
    background: @white;

    &:hover {
      background: #F2F5FF
    }

    &:focus {
      box-shadow: 0 1px 2px @blue-50, 0 0 0 3px @blue-50;
      border-radius: 6px;
    }

    &:active {
      background: @blue-25;
      border: 1px solid @blue-400;
      box-shadow: 0 1px 2px rgba(45, 54, 67, 0.05);
    }

    &:disabled {
      background: @white;
      border: 1px solid @blue-100;
      cursor: not-allowed;
    }
  }
}

.btn-sm:extend(.btn) {
  .paragraph-medium-5();
}


.btn-lg:extend(.btn) {
  .paragraph-medium-3();
}