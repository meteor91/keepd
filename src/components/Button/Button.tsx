import React from 'react';
import clsx from 'clsx';

import style from './Button.module.less';

type Variant = 'primary' | 'secondary' | 'tertiary';
export interface ButtonProps {
    label: string;
    disabled?: boolean;
    onClick?: (event?: React.MouseEvent<HTMLButtonElement>) => void;
    variant?: Variant;
}

export const Button: React.FC<ButtonProps> = (props) => {
    const {
        label,
        variant,
        disabled,
        onClick,
    } = props;

    return (
        <button
            onClick={onClick}
            className={clsx(style.btn, getButtonVariantClassName(variant))}
            disabled={disabled}
        >
            {label}
        </button>
    );
};

const getButtonVariantClassName = (variant?: Variant): string | null => {
    switch (variant) {
        case 'secondary':
            return style['btn-secondary'];
        default:
            return null;
    }
};
