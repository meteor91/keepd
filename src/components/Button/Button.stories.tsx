import React from 'react';
import type { Meta, StoryObj } from '@storybook/react';

import { Button } from './Button';

const meta: Meta<typeof Button> = {
    /* ðŸ‘‡ The title prop is optional.
   * See https://storybook.js.org/docs/react/configure/overview#configure-story-loading
   * to learn how to generate automatic titles
   */
    title: 'Button',
    component: Button,
    argTypes: {
        size: {
            control: { type: 'radio' },
            options: ['sm', 'md', 'lg'],
        },
        variant: {
            control: { type: 'radio' },
            options: ['primary', 'secondary'],
        },
        label: {
            control: 'text',
        },
        disabled: { control: 'boolean' },
    },
    args: {
        label: 'Button',
        size: 'md',
        variant: 'primary',
        disabled: false,
    },
};

export default meta;
type Story = StoryObj<typeof Button>;

/*
 *ðŸ‘‡ Render functions are a framework specific feature to allow you control on how the component renders.
 * See https://storybook.js.org/docs/react/api/csf
 * to learn how to use render functions.
 */
export const Basic: Story = {
    render: (args) => <Button {...args} />,
};
