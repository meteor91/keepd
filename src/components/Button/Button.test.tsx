import React from "react";
import { render, fireEvent } from "@testing-library/react";

import {Button} from "./Button";

describe("Button", () => {
  it('should render correctly', () => {
    const { container } = render(<Button label="test" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly disabled state', () => {
    const { container } = render(<Button label="test" disabled />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should call onClick only once when clicked', () => {
    const onClickMock = jest.fn();
    const { getByText } = render(<Button label="test" onClick={onClickMock}/>);
    const button = getByText('test');

    fireEvent.click(button);

    expect(onClickMock).toHaveBeenCalledTimes(1);
  });
});