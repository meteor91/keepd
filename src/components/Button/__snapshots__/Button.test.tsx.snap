// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render correctly 1`] = `
<button
  class="btn"
>
  test
</button>
`;

exports[`Button should render correctly disabled state 1`] = `
<button
  class="btn"
  disabled=""
>
  test
</button>
`;
